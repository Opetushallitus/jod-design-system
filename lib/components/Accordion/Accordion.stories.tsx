import type { StoryObj } from '@storybook/react-vite';
import type { TitledMeta } from '../../utils';
import { Accordion } from './Accordion';

const meta = {
  title: 'Content/Accordion',
  component: Accordion,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Accordion>;

export default meta;

type Story = StoryObj<typeof meta>;

// Decorator for preventing the preview size change when the content is toggled
const decorators: Story['decorators'] = [
  (Story) => (
    <div className="ds:min-h-[80px]">
      <Story />
    </div>
  ),
];

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-6882',
  },
};

export const Default: Story = {
  decorators,
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Accordion component for hiding content.',
      },
    },
  },
  args: {
    title: 'Title',
    children: 'Content',
    lang: 'en',
  },
};

export const TitleComponent: Story = {
  decorators,
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Accordion component with a custom component as a title.',
      },
    },
  },
  argTypes: {
    title: {
      control: false,
    },
  },
  args: {
    title: <div className="ds:text-heading-1 ds:font-arial ds:text-alert ds:italic ds:tracking-widest">Title</div>,
    titleText: 'Title',
    children: 'Content',
    lang: 'en',
  },
};

export const WithUnderline: Story = {
  decorators,
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Accordion component with underlined title.',
      },
    },
  },
  args: {
    title: 'Title',
    children: 'Content',
    underline: true,
    lang: 'en',
  },
};

export const WithAsync: Story = {
  decorators,
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Accordion component that shows a loading spinner while fetching data asynchronously.',
      },
    },
  },
  args: {
    title: 'Title',
    children: 'Content',
    underline: true,
    lang: 'en',
    initialState: false,
    fetchData: async () => new Promise((resolve) => setTimeout(resolve, 5000)),
  },
};

export const GrayClosedByDefault: Story = {
  decorators,
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Accordion that is closed by default and is gray variant',
      },
    },
  },
  args: {
    title: 'Title',
    children: <div className="ds:bg-bg-gray-2 ds:px-5 ds:pb-3">{'Content'}</div>,
    lang: 'en',
    initialState: false,
    className: 'ds:bg-bg-gray-2 ds:p-3',
  },
};
