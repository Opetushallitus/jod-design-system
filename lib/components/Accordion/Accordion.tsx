import React from 'react';
import { MdExpandLess, MdExpandMore } from 'react-icons/md';
import { cx } from '../../cva';

type TitleProps =
  | {
      title: React.ReactNode;
      titleText: string;
    }
  | {
      title: string;
      titleText?: never;
    };

type AccordionProps = {
  title: React.ReactNode | string;
  children: React.ReactNode;
  lang: string;
  underline?: boolean;
  intialState?: boolean;
} & TitleProps;

const Caret = ({ isOpen }: { isOpen: boolean }) => (
  <span className="ds:text-black ds:group-hover:text-accent!" aria-hidden>
    {isOpen ? <MdExpandLess size={24} /> : <MdExpandMore size={24} />}
  </span>
);

export const Accordion = ({ title, titleText, children, lang, underline, intialState = true }: AccordionProps) => {
  const [isOpen, setIsOpen] = React.useState(intialState);
  const toggleOpen = () => setIsOpen(!isOpen);
  const isTitleValidElement = React.isValidElement(title);
  const wrapperClassnames = cx(
    'ds:cursor-pointer ds:flex ds:w-full ds:items-center ds:justify-between ds:gap-x-4 ds:mb-2 ds:group-hover:text-accent!',
    {
      'ds:border-b ds:border-border-gray': underline,
    },
  );

  // Reset the state when the children change
  React.useEffect(() => {
    setIsOpen(intialState);
  }, [children, intialState]);

  return (
    <>
      {isTitleValidElement ? (
        <div className={wrapperClassnames}>
          {title}
          <button
            aria-label={titleText}
            aria-expanded={isOpen}
            onClick={toggleOpen}
            className="ds:cursor-pointer ds:flex"
          >
            <Caret isOpen={isOpen} />
          </button>
        </div>
      ) : (
        <div className="ds:group">
          <button aria-expanded={isOpen} onClick={toggleOpen} className={wrapperClassnames}>
            <span
              className="ds:mr-5 ds:w-full ds:text-left ds:hyphens-auto ds:text-heading-3 ds:group-hover:underline"
              lang={lang}
            >
              {title}
            </span>
            <Caret isOpen={isOpen} />
          </button>
        </div>
      )}
      {isOpen && children}
    </>
  );
};
