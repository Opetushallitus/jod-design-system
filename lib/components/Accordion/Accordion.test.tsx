import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Accordion } from './Accordion';

describe('Accordion', () => {
  const title = 'Accordion Title';
  const lang = 'en';

  test('renders accordion with open state by default', () => {
    const { container } = render(
      <Accordion title={title} lang={lang}>
        <div>Default content</div>
      </Accordion>,
    );

    const accordionTitle = screen.getByText(title);
    const expandButton = screen.getByRole('button');
    const accordionContent = screen.queryByText('Default content');

    expect(accordionTitle).toBeInTheDocument();
    expect(expandButton).toBeInTheDocument();
    expect(accordionContent).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test('closes accordion when expand button is clicked', () => {
    render(
      <Accordion title={title} lang={lang}>
        <div>Content to be gone</div>
      </Accordion>,
    );

    const expandButton = screen.getByRole('button');
    fireEvent.click(expandButton);

    const accordionContent = screen.queryByText('Content to be gone');
    expect(accordionContent).not.toBeInTheDocument();
  });
});
