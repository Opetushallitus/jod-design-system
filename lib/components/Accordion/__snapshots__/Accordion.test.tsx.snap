// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion > renders accordion with open state by default 1`] = `
<div
  class="ds:group"
>
  <button
    aria-expanded="true"
    class="ds:cursor-pointer ds:flex ds:w-full ds:items-center ds:justify-between ds:gap-x-4 ds:mb-2 ds:group-hover:text-accent!"
  >
    <span
      class="ds:mr-5 ds:w-full ds:text-left ds:hyphens-auto ds:text-heading-3 ds:group-hover:underline"
      lang="en"
    >
      Accordion Title
    </span>
    <span
      aria-hidden="true"
      class="ds:text-black ds:group-hover:text-accent!"
    >
      <svg
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"
        />
      </svg>
    </span>
  </button>
</div>
`;
